src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+    is no penalty for doing so. <gdImageAvifEx> does not close the file;
+    your code must do so.
+
+  Variants:
+
+    <gdImageAvifEx> writes the image to a file, encoding with the default quality and speed.
+
+    <gdImageAvifPtrEx> stores the image in RAM.
+
+    <gdImageAvifPtr> stores the image in RAM, encoding with the default quality and speed.
+
+    <gdImageAvifCtx> stores the image using a <gdIOCtx> struct.
+
+
+  Parameters:
+
+    im      - The image to save.
+    outFile - The FILE pointer to write to.
+    quality - Compression quality (0-100). 0 is lowest-quality, 100 is highest.
+    speed   - The speed of compression (0-10). 0 is slowest, 10 is fastest.
+
+  Notes on parameters: 
+    quality - If quality = -1, we use a default quality as defined in DEFAULT_QUALITY.
+      For information on how we convert this quality to libavif's quantity param, see <quality2Quantizer>
+
+    speed - At slower speeds, encoding may be quite slow. Use judiciously.
+
+  Returns:
+
+    Nothing.
+*/
+
+BGD_DECLARE(void) gdImageAvifEx (gdImagePtr im, FILE *outFile, int quality, int speed)
+{
+	gdIOCtx *out = gdNewFileCtx(outFile);
+
+	if (out == NULL) {
+		return;
+	}
+
+	gdImageAvifCtx(im, out, quality, speed);
+	out->gd_free(out);
+}
+
+BGD_DECLARE(void) gdImageAvif (gdImagePtr im, FILE *outFile)
+{
+	return gdImageAvifEx(im, outFile, DEFAULT_QUALITY, AVIF_SPEED_DEFAULT);
+}
+
+BGD_DECLARE(void *) gdImageAvifPtrEx (gdImagePtr im, int *size, int quality, int speed)
+{
+	void *rv;
+	gdIOCtx *out = gdNewDynamicCtx(NEW_DYNAMIC_CTX_SIZE, NULL);
+
+	if (out == NULL) {
+		return NULL;
+	}
+
+	if (_gdImageAvifCtx(im, out, quality, speed)) {
+		rv = NULL;
+	} else {
+		rv = gdDPExtractData(out, size);
+	}
+
+	out->gd_free(out);
+	return rv;
+}
+
+BGD_DECLARE(void *) gdImageAvifPtr (gdImagePtr im, int *size)
+{
+	return gdImageAvifPtrEx(im, size, DEFAULT_QUALITY, AVIF_SPEED_DEFAULT);
+}
+
+
+BGD_DECLARE(void) gdImageAvifCtx (gdImagePtr im, gdIOCtx *outfile, int quality, int speed)
+{
+  _gdImageAvifCtx(im, outfile, quality, speed);
+}
+
+/* 
+   We need this underscored function because gdImageAvifCtx() can't return anything.
+   And our functions that operate on a memory buffer need to know whether the encoding has succeeded.
+
+   If we're passed the DEFAULT_QUALITY of -1, set the quantizer params to DEFAULT_QUANTIZER
+
+   This function returns 0 on success, or 1 on failure.
+ */
+static avifBool _gdImageAvifCtx (gdImagePtr im, gdIOCtx *outfile, int quality, int speed)
+{
+  avifResult result;
+  avifRGBImage rgb;
+  avifRWData avifOutput = AVIF_DATA_EMPTY;
+  avifBool failed = AVIF_FALSE;
+
+  uint32_t val;
+  uint8_t *p;
+  uint8_t a;
+  int x, y;
+
+	if (im == NULL) {
+		return 1;
+	}
+
+	if (!gdImageTrueColor(im)) {
+		gd_error("avif doesn't support palette images");
+		return 1;
+	}
+
+  avifImage *avifIm = avifImageCreate(gdImageSX(im), gdImageSY(im), 8, DEFAULT_CHROMA_SUBSAMPLING);
```

Comment:
I would use AVIF_PIXEL_FORMAT_YUV422 above 90% quality.

Reply:
That's an interesting idea...

@wantehchang, do you have an opinion here? The idea intrigues me... my only concern is that it may make our `quality` param more obtuse than it already is.

Reply:
The YUV 4:2:2 format is not commonly used. You can use AVIF_PIXEL_FORMAT_YUV444 if you don't want chroma subsampling when the quality level is above 90%.

Reply:
Did this!